core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala

```diff
@@ -291,6 +297,16 @@ private[spark] trait SparkListenerInterface {
   def onBlockUpdated(blockUpdated: SparkListenerBlockUpdated): Unit
 
   /**
+   * Called when a speculative task is submitted
+   */
+  def onSpeculativeTaskSubmitted(speculativeTask: SparkListenerSpeculativeTaskSubmitted): Unit
+
+  /**
+   * Called when an extra executor is needed
+   */
+  def onExtraExecutorNeeded(): Unit
```

Comment:
I'm hesitant to add this. `SparkListener` is public and should be the interface to listen to various Spark internal events, and do whatever they want. However, `onExtraExecutorNeeded` sounds like something Spark asks the listener to do, which is against the pattern.

In another word, `onExtraExecutorNeeded` looks specific to the executor allocation manager, but not a general spark event.